### PR TITLE
[SPARK-40530][SQL] Add error-related developer APIs

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

Third-party Spark plugins may define their own errors using the same framework as Spark: put error definition in json files. This PR moves some error-related code to a new file and marks them as developer APIs, so that others can reuse them instead of writing its own json reader.

### Why are the changes needed?

make it easier for Spark plugins to define errors.

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

existing tests

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/SparkThrowableHelper.scala [hunk 1]
```diff
@@ -69,37 +33,8 @@ object ErrorMessageFormat extends Enumeration {
...
-  val errorClassesUrl: URL =
-    Utils.getSparkClassLoader.getResource("error/error-classes.json")
-  val errorClassToInfoMap: SortedMap[String, ErrorInfo] = {
-    val mapper: JsonMapper = JsonMapper.builder()
-      .addModule(DefaultScalaModule)
-      .build()
-    mapper.readValue(errorClassesUrl, new TypeReference[SortedMap[String, ErrorInfo]]() {})
-  }
-
-  def getParameterNames(errorClass: String, errorSubCLass: String): Array[String] = {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/SparkThrowableHelper.scala:81]: not related to this PR, but I found this method is not called anywhere.

**Comment by cloud-fan** [comment]
cc @MaxGekk @viirya

File: core/src/main/scala/org/apache/spark/SparkThrowableHelper.scala [hunk 2]
```diff
@@ -69,37 +31,8 @@ object ErrorMessageFormat extends Enumeration {
...
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/SparkThrowableHelper.scala:81]: this method is called no where.

File: core/src/main/scala/org/apache/spark/ErrorClassesJSONReader.scala [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+import com.fasterxml.jackson.module.scala.DefaultScalaModule
+import org.apache.commons.text.StringSubstitutor
+
+import org.apache.spark.annotation.DeveloperApi
+
+/**
+ * A reader to load error information from one or more JSON files. Note that, if one error appears
+ * in more than one JSON files, the latter wins.
+ */
+@DeveloperApi
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/ErrorClassesJSONReader.scala:37]: As a developer API, do we need to briefly describe the error JSON format here?

File: core/src/main/scala/org/apache/spark/ErrorClassesJSONReader.scala [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+  def getErrorMessage(errorClass: String, messageParameters: Map[String, String]): String = {
+    val errorClasses = errorClass.split("\\.")
+    assert(errorClasses.length == 1 || errorClasses.length == 2)
+
+    val mainErrorClass = errorClasses.head
+    val subErrorClass = errorClasses.tail.headOption
+    val errorInfo = errorInfoMap.getOrElse(
+      mainErrorClass,
+      throw SparkException.internalError(s"Cannot find error class '$errorClass'"))
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/ErrorClassesJSONReader.scala:59]: Here should be `mainErrorClass`? in the internal error message?

File: core/src/main/scala/org/apache/spark/ErrorClassesJSONReader.scala [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    assert(errorInfo.subClass.isDefined == subErrorClass.isDefined)
+
+    val finalFormat = if (subErrorClass.isEmpty) {
+      errorInfo.messageFormat
+    } else {
+      val errorSubInfo = errorInfo.subClass.get.getOrElse(
+        subErrorClass.get,
+        throw SparkException.internalError(s"Cannot find sub error class '${subErrorClass.get}'"))
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/ErrorClassesJSONReader.scala:67]: Why not keep IllegalArgumentException?

File: core/src/test/scala/org/apache/spark/SparkThrowableSuite.scala [hunk 6]
```diff
@@ -137,22 +139,22 @@ class SparkThrowableSuite extends SparkFunSuite {
...
+    mapper.writeValue(tmpFile, errorReader.errorInfoMap)
     val rereadErrorClassToInfoMap = mapper.readValue(
       tmpFile, new TypeReference[Map[String, ErrorInfo]]() {})
-    assert(rereadErrorClassToInfoMap == errorClassToInfoMap)
+    assert(rereadErrorClassToInfoMap == errorReader.errorInfoMap)
   }
 
   test("Check if error class is missing") {
-    val ex1 = intercept[IllegalArgumentException] {
+    val ex1 = intercept[SparkException] {
```
Comment (viirya) [inline core/src/test/scala/org/apache/spark/SparkThrowableSuite.scala:149]: Related to https://github.com/apache/spark/pull/37969/files#r978093752.

**Review by viirya — COMMENTED** [review]
Looks good overall. Just a few minor comments.

File: core/src/main/scala/org/apache/spark/ErrorClassesJSONReader.scala (hunk 4 above)
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/ErrorClassesJSONReader.scala:59]: I think it's better to include more information and put the full error class (main + sub)

File: core/src/main/scala/org/apache/spark/SparkThrowableHelper.scala [hunk 7]
```diff
@@ -178,9 +90,7 @@ private[spark] object SparkThrowableHelper {
           val errorSubClass = e.getErrorSubClass
           if (errorSubClass != null) g.writeStringField("errorSubClass", errorSubClass)
           if (format == STANDARD) {
-            val errorInfo = errorClassToInfoMap.getOrElse(errorClass,
-              throw SparkException.internalError(s"Cannot find the error class '$errorClass'"))
-            g.writeStringField("message", errorInfo.messageFormat)
+            g.writeStringField("message", e.getMessage)
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/SparkThrowableHelper.scala:94]: Makes sense, we should allow the caller side to do substitution as they want.

File: core/src/main/scala/org/apache/spark/ErrorClassesJSONReader.scala (hunk 3 above)
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/ErrorClassesJSONReader.scala:38]: Added the reference to the readme file of the error framework.

File: core/src/test/scala/org/apache/spark/SparkThrowableSuite.scala [hunk 8]
```diff
@@ -321,4 +323,22 @@ class SparkThrowableSuite extends SparkFunSuite {
...
+        """
+          |{
+          |  "DIVIDE_BY_ZERO" : {
+          |    "message" : [
+          |      "abc"
+          |    ]
+          |  }
+          |}
+          |""".stripMargin)
+      val reader = new ErrorClassesJsonReader(Seq(errorJsonFilePath.toUri.toURL, json.toURL))
```
Comment (MaxGekk) [inline core/src/test/scala/org/apache/spark/SparkThrowableSuite.scala:340]: Do you use the existing JSON file? How can you guarantee that one JSON overrides the standard one? Maybe you will add small one for testing to check that it overrides error-classes.json?

Comment (cloud-fan) [reply core/src/test/scala/org/apache/spark/SparkThrowableSuite.scala:340]: This test uses 2 JSON files: the existing official one and a temporary one which was written in this test.

**Comment by cloud-fan** [comment]
thanks for review, merging to master!

**Pull request closed without merging**